### PR TITLE
feat: load tag from metastore

### PR DESCRIPTION
**Pull request by jczhong84**
Allow loading tags from metastore
 - By default, it will not load tags from metastore. 
 - To enable it
     - Update metastore loader's `loader_config` to have `MetadataType.TAG` not equal to ` MetadataMode.WRITE_LOCAL`. 
     - Update metastore loader's `get_table_and_columns` function to return tags as `table.tags`
 - If enabled loading tags from metastore, tags can't be added or updated through Querybook UI.
 - All the tags synced from metastore will have `tag.admin = True`
 - Added a new meta field `meta.type` to indicate a tag type. If a tag has a type, it will be rendered along with the tag name as a tag group. see below screenshot as an example
 
image

### CODE REVIEW

File: querybook/webapp/ui/Tag/HoverIconTag.tsx [hunk 1]
```diff
@@ -4,28 +4,73 @@ import React from 'react';
...
+        if (tagType) {
+            return (
+                <span
+                    aria-label={tooltip}
+                    data-balloon-pos={tooltipPos}
+                    onClick={onClick}
+                    ref={ref}
+                    className={className}
+                >
+                    <TagGroup>
```
Comment (czgu) [inline querybook/webapp/ui/Tag/HoverIconTag.tsx:54]: i just realized TagGroup is just a div wrapper, can you convert it into styled components and pass in the props above to there instead?

File: querybook/webapp/ui/Tag/HoverIconTag.tsx [hunk 2]
```diff
@@ -4,28 +4,73 @@ import React from 'react';
...
+    ) => {
+        const hoverDOM = iconOnHover ? (
+            <div className="HoverIconTag-hover" onClick={onIconHoverClick}>
+                <Icon name={iconOnHover} />
+            </div>
+        ) : null;
+
+        const className = clsx(tagProps['className'], 'HoverIconTag');
+
+        const { tooltip, tooltipPos, color, mini, onClick } = tagProps;
```
Comment (czgu) [inline querybook/webapp/ui/Tag/HoverIconTag.tsx:39]: move this into the if(tagType) part

File: querybook/webapp/ui/Tag/HoverIconTag.tsx [hunk 3]
```diff
@@ -4,28 +4,73 @@ import React from 'react';
...
+                    onClick={onClick}
+                    ref={ref}
+                    className={className}
+                >
+                    <TagGroup>
+                        <Tag mini={mini}>
+                            {iconDOM}
+                            <span>{tagType}</span>
+                        </Tag>
+                        <Tag mini={mini} highlighted color={color}>
```
Comment (czgu) [inline querybook/webapp/ui/Tag/HoverIconTag.tsx:59]: you still need to forward whatever is left of tagprops into this?

File: querybook/server/lib/metastore/base_metastore_loader.py [hunk 4]
```diff
@@ -380,6 +337,14 @@ def _create_table_table(
                     commit=False,
                     session=session,
                 )
+
+            # create tags if the metastore is configured to sync tags
+            if self.loader_config.load_metadata(MetadataType.TAG):
```
[the remark this answers is not in the document]
Comment (jczhong84) [reply querybook/server/lib/metastore/base_metastore_loader.py:342]: I think it's needed b/c user can always refresh/sync a specific table from metastore.

File: querybook/server/logic/tag.py [hunk 5]
```diff
@@ -89,3 +89,43 @@ def delete_tag_from_table(
...
+        # filter out properties with none values
+        meta = {k: v for k, v in meta.items() if v is not None}
+
+        # update or create a new tag if not exist
+        create_or_update_tag(
+            tag_name=tag.name, meta=meta, commit=commit, session=session
+        )
+
+        # add a new tag_item to associate with the table
+        tag_item = TagItem.get(table_id=table_id, tag_name=tag.name, session=session)
```
[the remark this answers is not in the document]
Comment (jczhong84) [reply querybook/server/logic/tag.py:121]: you're right

File: querybook/server/const/metastore.py [hunk 6]
```diff
@@ -37,5 +38,12 @@ class MetastoreLoaderConfig:
     def __init__(self, config: dict[MetadataType, MetadataMode]):
         self._config = {**self._default_config, **config}
 
+    def can_load_external_metadata(self, metadataType: MetadataType) -> bool:
+        """Check if the given metadata type will be loaded from metastore"""
+        return (
+            self._config.get(metadataType, MetadataMode.WRITE_LOCAL)
+            != MetadataMode.WRITE_LOCAL
```
Comment (czgu) [inline querybook/server/const/metastore.py:45]: minor nit: this should be `in (WRITE_LOCAL, WRITE_BACK)`

File: querybook/server/lib/metastore/base_metastore_loader.py [hunk 7]
```diff
@@ -384,6 +338,14 @@ def _create_table_table(
                     commit=False,
                     session=session,
                 )
+
+            # create tags if the metastore is configured to sync tags
+            if self.loader_config.can_load_external_metadata(MetadataType.TAG):
+                create_table_tags(
+                    table_id=table_id,
+                    tags=table.tags,
```
Comment (czgu) [inline querybook/server/lib/metastore/base_metastore_loader.py:346]: commit=False?

**Review by czgu — APPROVED** [review]

**Pull request merged**